operator/v1/types_baremetal.go

```diff
@@ -0,0 +1,110 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// Metal3Provisioning contains configuration used by the Provisioning
+// service (Ironic) to provision baremetal hosts.
+//
+// Metal3Provisioning is created by the Openshift installer using admin
+// or user provided information about the provisioning network and the NIC
+// on the server that can be used to PXE boot it.
+//
+// This CR is a singleton, created by the installer and currently only
+// consumed by the machine-api-operator to bring up and update containers
+// in a metal3 cluster.
+//
+type Metal3Provisioning struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the
+	// Metal3Provisioning.
+	Spec Metal3ProvisioningSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the
+	// Metal3Provisioning.
+	Status Metal3ProvisioningStatus `json:"status,omitempty"`
+}
+
+// ProvisioningDHCP represents just the configuration required to fully
+// identify the way DHCP would be handled during baremetal host bringup.
+//
+// DHCP services could be provided external to the metal3 cluster, in
+// which case, IP address assignment for the baremetal servers should
+// happen via this external DHCP server and not via a DHCP server started
+// within the metal3 cluster.
+// If IP address assignment needs to happen via the DHCP server within the
+// metal3 cluster, then the CR needs to contain the DHCP address range that
+// this internal DHCP server needs to use.
+//
+type ProvisioningDHCP struct {
+	// ManagementState within the OperatorStatus needs to be set to
+	// indicate if the DHCP server is internal or external to the
+	// metal3 cluster. ManagementState set to "Removed" indicates
+	// that the DHCP server is outside the metal3 cluster. And a
+	// value of "Managed" indicates that the DHCP services are
+	// managed within the metal3 cluster.
+	// The other fields of OperatorStatus retain their existing
+	// semantics.
+	OperatorStatus `json:",inline"`
```

Comment:
sounds like you want a discrimated union.  a string indicates the type then separate stanzas for the details.  @damemi probably has a link.

Reply:
I think we had considered that during the enhancement request. Then we settled upon (re)using ManagementState and now expanded to use the OperatorStatus based on the fact that we could use the other fields in OperatorStatus also (and based comments received in an earlier version).

Reply:
I made a mistake here. This should be OperatorSpec.